model/test_results.go

```diff
@@ -173,50 +171,38 @@ func (t *TestResults) Append(ctx context.Context, results []TestResult) error {
 		true,
 	)
 	if err != nil {
-		return errors.Wrap(err, "problem creating bucket")
-	}
-
-	resultChan := make(chan *TestResult, len(results))
-	for i := 0; i < len(results); i++ {
-		resultChan <- &results[i]
-	}
-	close(resultChan)
-	var wg sync.WaitGroup
-	catcher := grip.NewBasicCatcher()
-	for i := 0; i < runtime.NumCPU(); i++ {
-		wg.Add(1)
-		go func() {
-			defer func() {
-				catcher.Add(recovery.HandlePanicWithError(recover(), nil, "test results append"))
-				wg.Done()
-			}()
-
-			for result := range resultChan {
-				data, err := bson.Marshal(result)
-				if err != nil {
-					// We should continue on marhsalling
-					// errors so we can get as many test
-					// results in s3 as possible.
-					catcher.Wrapf(err, "problem marshalling bson for test result '%s'", result.TestName)
-					continue
-				}
-
-				if err := bucket.Put(ctx, result.TestName, bytes.NewReader(data)); err != nil {
-					catcher.Wrapf(err, "problem uploading test result '%s' to bucket", result.TestName)
-					return
-				}
-			}
-		}()
-	}
-	wg.Wait()
-
-	return catcher.Resolve()
+		return errors.Wrap(err, "creating bucket")
+	}
+
+	var allResults testResultsDoc
+	r, err := bucket.Get(ctx, testResultsCollection)
+	if err != nil && !pail.IsKeyNotFoundError(err) {
+		return errors.Wrap(err, "getting uploaded test results")
+	}
+	if r != nil {
+		data, err := ioutil.ReadAll(r)
+		if err != nil {
+			return errors.Wrap(err, "reading uploaded test results")
+		}
+
+		if err = bson.Unmarshal(data, &allResults); err != nil {
+			return errors.Wrap(err, "unmarshalling uploaded test results")
+		}
+	}
+	allResults.Results = append(allResults.Results, results...)
+
+	data, err := bson.Marshal(&allResults)
+	if err = bucket.Put(ctx, testResultsCollection, bytes.NewReader(data)); err != nil {
+		return errors.Wrap(err, "uploading test results")
+	}
+
+	return nil
 }
 
 // Download returns a TestResult slice with the corresponding results stored in
 // the offline blob storage. The TestResults should be populated and the
 // environment should not be nil.
-func (t *TestResults) Download(ctx context.Context) (TestResultsIterator, error) {
+func (t *TestResults) Download(ctx context.Context) ([]TestResult, error) {
```

Comment:
Yeah, since the results should cap around 10MB (most being much smaller) it will actually be very efficient. The current bottleneck is making thousands of individual requests to s3 when fetching each test result stored separately. We are not currently paginating, but that may occur in a follow-up ticket. I need to continue working with UI as they develop new features.